### PR TITLE
[Feature] Storage API + Metrics API

### LINKED ISSUE #52: Missions and Achievements
**Feature Request**

Some Achievements / Missions to consider:

Completionist: Complete all missions and side objectives in the game.

Speedrunner: Complete a mission or level within a certain time limit.

Master Strategist: Complete a mission using only specific units or abilities, demonstrating advanced strategic prowess.

Untouchable: Complete a mission without taking any damage or losing any units.

### PR DESCRIPTION
**Pull request by sp4ce-cowboy**
### Summary
This PR adds the entire Metrics API + Storage API. Added both together given the close link between them (i.e. Storage is needed to see if the Metrics work properly).

#### MetricsAPI
- Statistics sub-API: Gathers information using an event-driven architecture, and liaises with Storage.
- Inference sub-API: Custom contexts to make sense of statistics gathered above. 
    - Achievements: Permanent achievements completed throughout the lifetime of the app
    - Missions: Objectives completed within a given game sequence only
    - Ranking: Experience points and rank can be determined from Statistics (architecturally scalable to be used for level-based player matching or level-based perks)

#### StorageAPI
- Local & Remote: All storage is stored both locally and remotely in firebase, regardless of whether user is logged in by using one-time instantiated UUID that is stored separately and used to distinguish between different devices that ping firebase.
- Metadata: The above mentioned one-time instantiation is stored within a Metadata file identifies the user and simultaneously keeps track of latest update date signature. Metadata is also stored locally and remotely. 
- Data collision resolution: 2 types of data collision resolution implemented (Merge both or keep latest only) using the above metadata and some manual merging.
- Data synchronization: Above-mentioned collision also takes into account cases where a user is playing locally, and then decides to log in. All progress can be merged, and kept in sync.

#### Pending & Concerns
- Singleton pattern eliminated, but given the nature of Firebase/local storage, only utility classes with static methods are used. This prevents data from existing in two places at once (i.e. inside game memory and local storage) but it's a bit too close to singleton pattern, might need to double confirm.
- Achievements, Missions need a UI so that the user can view them. (Ideally can move to a new PR)
- Data sync and remote storage could use some more testing to make sure they work with user login.

### PR Status
Ready for review. UIView for Achievements and missions to be move to another PR. Some commented code is retained just in case the newer StatisticsUpdateActors fail and we need to revert back (which we can only tell after some more testing)

Closes #52

### CODE REVIEW

File: TowerForge/TowerForge/Commons/Protocols/Double+Extensions.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
+//
+//  Double+Extensions.swift
+//  TowerForge
+//
+//  Created by Rubesh on 14/4/24.
+//
+
+import Foundation
+
+extension Double {
```
Comment (zheng-ze) [inline TowerForge/TowerForge/Commons/Protocols/Double+Extensions.swift:10]: I think this file should be in the extensions folder

File: TowerForge/TowerForge/Metrics/Statistics/Statistic.swift [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                                                              currentValue: currentValue) else {
+
+            throw DecodingError.dataCorruptedError(forKey: .statisticName,
+                                                   in: container,
+                                                   debugDescription: "Cannot instantiate Statistic of type \(typeName)")
+        }
+
+        self = instance
+    }*/
+}
```
Comment (zheng-ze) [inline TowerForge/TowerForge/Metrics/Statistics/Statistic.swift:159]: I think you accidentally committed this.

File: TowerForge/TowerForge/Storage/StorageManager.swift [hunk 3]
```diff
@@ -1,121 +1,109 @@
...
-            let data = try encoder.encode(storedDatabase)
-            let folderURL = try Self.createFolderIfNeeded(folderName: Self.folderName)
-            let fileURL = folderURL.appendingPathComponent(fileNameCombined)
-            try data.write(to: fileURL)
-            Logger.log("Saved Storage at: \(fileURL.path)")
-        } catch {
-            Logger.log("Error saving Database: \(error)")
-        }
+    func onLogin() {
+
```
Comment (zheng-ze) [inline TowerForge/TowerForge/Storage/StorageManager.swift:27]: I think you should perform a sync on load?

**Review by zheng-ze — APPROVED** [review]
I think generally this looks ok.

File: TowerForge/TowerForge/Commons/Protocols/Double+Extensions.swift (hunk 1 above)
Comment (sp4ce-cowboy) [reply TowerForge/TowerForge/Commons/Protocols/Double+Extensions.swift]: Oh nice thanks for letting me know, that seems to be the issue. It's fixed now.

**Review by zheng-ze — APPROVED** [review]
Looks fine but I feel that the retrieval of stats from events is a bit hardcoded and there is some commented code that needs to be cleaned up.

File: TowerForge/TowerForge/GameModule/Events/GameEvents/DamageEvent.swift [hunk 4]
```diff
@@ -25,6 +25,13 @@ struct DamageEvent: TFEvent {
         if let healthSystem = target.system(ofType: HealthSystem.self) {
             healthSystem.modifyHealth(for: entityId, with: -damage)
         }
+
+        if let statsSystem = target.system(ofType: StatisticSystem.self) {
+            if player != .ownPlayer {
+                statsSystem.notify(for: self)
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (sp4ce-cowboy) [reply TowerForge/TowerForge/GameModule/Events/GameEvents/DamageEvent.swift:34]: Thanks for noticing this, I forgot to remove the player check here. I've implemented something similar which removes the hard coding:

```swift
        let eventUpdateClosure: (Statistic, DamageEvent?) -> Void = { statistic, event in
            guard let event = event, event.player != .ownPlayer else {
                return
            }
            statistic.updateCurrentValue(by: Double(event.damage))
            Logger.log("Updating statistic with event detail: \(String(describing: event))", self)
        }
```

This code snippet is from the TotalDamageDealtStatistic, the StatisticUpdateActor (i.e. eventUpdateClosure) related to this Statistic now takes in an Event as opposed to just EventType from the prior implementation (which had the check inside the event itself) and the parsing is done by the StatsEngine when it calls to execute the above closure.

File: TowerForge/TowerForge/Metrics/Statistics/Implemented/TotalDamageDealtStatistic.swift [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /*func getStatisticUpdateLinks() -> StatisticUpdateLinkDatabase {
+        let eventType = TFEventTypeWrapper(type: DamageEvent.self)
+        let updateActor: StatisticUpdateActor = { statistic in statistic.updateCurrentValue(by: ) }
+        let eventUpdateDictionary = [eventType: updateActor]
+        let statsLink = StatisticUpdateLinkDatabase(statisticUpdateLinks: eventUpdateDictionary)
+
+        return statsLink
+    }*/
```
[the remark this answers is not in the document]
Comment (sp4ce-cowboy) [reply TowerForge/TowerForge/Metrics/Statistics/Implemented/TotalDamageDealtStatistic.swift]: Fixed, thanks.

**Review by zheng-ze — APPROVED** [review]

**Pull request merged**